core/types/deposit.go

```diff
@@ -0,0 +1,118 @@
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU Lesser General Public License for more details.
+//
+// You should have received a copy of the GNU Lesser General Public License
+// along with the go-ethereum library. If not, see <http://www.gnu.org/licenses/>.
+
+package types
+
+import (
+	"bytes"
+	"encoding/binary"
+
+	"github.com/ethereum/go-ethereum/accounts/abi"
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/common/hexutil"
+	"github.com/ethereum/go-ethereum/rlp"
+)
+
+//go:generate go run github.com/fjl/gencodec -type Deposit -field-override depositMarshaling -out gen_deposit_json.go
+
+// Deposit contians EIP-6110 deposit data.
+type Deposit struct {
+	PublicKey             [48]byte     `json:"pubkey"`                // public key of validator
+	WithdrawalCredentials common.Hash  `json:"withdrawalCredentials"` // beneficiary of the validator funds
+	Amount                uint64       `json:"amount"`                // deposit size in Gwei
+	Signature             [96]byte     `json:"signature"`             // signature over deposit msg
+	Index                 uint64       `json:"index"`                 // deposit count value
+}
+
+// field type overrides for gencodec
+type depositMarshaling struct {
+	PublicKey             hexutil.Bytes
+	WithdrawalCredentials hexutil.Bytes
+	Amount                hexutil.Uint64
+	Signature             hexutil.Bytes
+	Index                 hexutil.Uint64
+}
+
+// field type overrides for abi upacking
+type depositUnpacking struct {
+	Pubkey                []byte
+	WithdrawalCredentials []byte
+	Amount                []byte
+	Signature             []byte
+	Index                 []byte
+}
+
+// Deposits implements DerivableList for requests.
+type Deposits []*Deposit
+
+// Len returns the length of s.
+func (s Deposits) Len() int { return len(s) }
+
+// EncodeIndex encodes the i'th deposit to s.
+func (s Deposits) EncodeIndex(i int, w *bytes.Buffer) {
+	rlp.Encode(w, s[i])
+}
+
+// Requests creates a deep copy of each deposit and returns a slice of Request
+// objects.
+func (s Deposits) Requests() (reqs Requests) {
+	for _, d := range s {
+		reqs = append(reqs, NewRequest(d))
+	}
+	return
+}
+
+var (
+	// DepositABI is an ABI instance of beacon chain deposit events.
+	DepositABI   = abi.ABI{Events: map[string]abi.Event{"DepositEvent": depositEvent}}
+	bytesT, _    = abi.NewType("bytes", "", nil)
+	depositEvent = abi.NewEvent("DepositEvent", "DepositEvent", false, abi.Arguments{
+		{Name: "pubkey", Type: bytesT, Indexed: false},
+		{Name: "withdrawal_credentials", Type: bytesT, Indexed: false},
+		{Name: "amount", Type: bytesT, Indexed: false},
+		{Name: "signature", Type: bytesT, Indexed: false},
+		{Name: "index", Type: bytesT, Indexed: false}},
+	)
+)
+
+// UnpackIntoDeposit unpacks a serialized DepositEvent.
+func UnpackIntoDeposit(data []byte) (*Deposit, error) {
+	var du depositUnpacking
```

Comment:
We need to hand-write this ABI decoder.

Reply:
done in [268e39b](https://github.com/ethereum/go-ethereum/pull/29431/commits/268e39b12eb7dcd5afed004c5d8bc0a271ef9a04)